### PR TITLE
67: Auto-inline non-main functions

### LINKED ISSUE #67: Auto-inline non-main functions
**Is your feature request related to a problem? Please describe.**
To allow for greater default optimizations, denoting functions as inline should allow for optimizing away entire programs.

**Describe the solution you'd like**
Automatically add `inline` to all non-`main` functions

**Describe alternatives you've considered**
Add an inline keyword. Inlining should be Daybreak's default, so a keyword for inlining is senseless.

**Additional context**
Euler problem 001 generates C code that, when compiled with the same flags but inlining the functions, allows for a simple return statement with the answer. Static analysis of Daybreak programs will be greatly aided by marking that all functions are able to be inlined.

### PR DESCRIPTION
**Pull request by JamesOrson**
Closes #67

### CODE REVIEW

**Review by brandonapol — COMMENTED** [review]
The -02 flag signals something to the compiler - this allows for GCC to optimize compiling. I assume this is to avoid using as much memory as the -0 flag? Are there other optimizations that might be better for the specific purposes of working with the C compiler at this point?

Is the goal of inlining the compiler functions similar to writing anonymous functions? Essentially, if I understand correctly, this update will allow your anonymous functions to compile more quickly and efficiently by using better flags to direct the C compiler, and it checks the OS (WIN32 vs WIN64 vs else) as well to modify whether the function is static. The value is ultimately saved as a boolean which you've added here as to whether or not it's an inline.

**Comment by JamesOrson** [comment]
So the -O2 flag optimizes more harshly than -O

Specifically, I am trying to accomplish "inlining" of all daybreak functions, meaning they can be directly placed in the binary and not incur the cost of a new stack frame (the way a function call does).

If you don't do either -O2 or -O3, inline functions cause a compiler linker error because C compilers are weird. ¯\_(ツ)_/¯

Now about `static inline` vs `inline`, Windows GCC is odd in that it's more strict about certain functions not being callable from an `inline` function unless it is also `static`. For example, `printf` is a static function so any `inline` function also has to be `static` on Windows for C.

This is some weird "in the weeds" stuff with C compilers

tl;dr This code change makes all daybreak functions compile as `inline` functions allowing compilers to do harsher optimizations on daybreak code. Because daybreak code is immutable and recursive, it optimized very well when inlined and much of the code is moved entirely to compile time. Other changes are specific to Windows being weird about certain C compilation rules.

Cool results, these changes made all my Euler problem code compile away completely. This means the optimizations worked exactly as intended, moving a huge amount of work into the compilation step itself: https://en.m.wikipedia.org/wiki/Compile-time_function_execution

**Comment by JamesOrson** [comment]
Now the last thing you asked was about a boolean being stored.

So the bool was just the return type of the `_should_inline` function I wrote, and the windows check is actually something done by the compiler that CONDITIONALLY compiles the code between the
```c
#if defined(WIN32) || defined(WIN64)
// this code is compiled if Windows 
#else
// this code is compiled if not Windows
#endif
```

**Comment by JamesOrson** [comment]
> Now the last thing you asked was about a boolean being stored.
> 
> So the bool was just the return type of the `_should_inline` function I wrote, and the windows check is actually something done by the compiler that CONDITIONALLY compiles the code between the
> ```c
> #if defined(WIN32) || defined(WIN64)
> // this code is compiled if Windows 
> #else
> // this code is compiled if not Windows
> #endif
> ```

More info on this: https://stackoverflow.com/a/5920028

File: src/include/generation/generate_c.h [hunk 1]
```diff
@@ -290,14 +293,22 @@ int generate_c_from_field_list(FILE* output_file, const struct FieldListNode* fi
...
 }
 
-int generate_c_from_function_declaration(
-	FILE* output_file,
-	const struct FunctionDeclarationNode* function_declaration
-) {
+int _generate_c_function_signature(FILE* output_file, const struct FunctionDeclarationNode* function_declaration, const bool is_inline) {
 	if (!function_declaration) {
-		LOG_ERROR("Failed to generate C code from FunctionDeclarationNode. NULL FunctionDeclarationNode.");
+		LOG_ERROR("Failed to generate common function declaration from FunctionDeclarationNode. NULL FunctionDeclarationNode.");
```
Comment (JamesOrson) [inline src/include/generation/generate_c.h:298]: Change this to mention "generate function signature"

Comment (JamesOrson) [reply src/include/generation/generate_c.h]: Fixed

**Review by brandonapol — APPROVED** [review]
Looks like you added some CI/CD stuff, monitoring with valgrind, the rest of this is still architecting the inlining process and error handling for it. What exactly is a C Function Signature? I take it this is Daybreak generating C code for part of compiling?

**Comment by JamesOrson** [comment]
> Looks like you added some CI/CD stuff, monitoring with valgrind, the rest of this is still architecting the inlining process and error handling for it. What exactly is a C Function Signature? I take it this is Daybreak generating C code for part of compiling?

- CI/CD got updated to have ANSI color output in Jenkins
- Just changed order of arguments to `apt` install of `valgrind`
- Yup, inlining daybreak functions and the error handling
- I decided to call the common thing between a function declaration and a function defintion in C a `function signature`
```c
// C function declaration
inline int double(int a);
// C function definition
int double(int a) {
  return a * 2;
}

// Both the declaration and definition have something in common which I have deemed a `signature`
/*
int double(int a)
*/
```
